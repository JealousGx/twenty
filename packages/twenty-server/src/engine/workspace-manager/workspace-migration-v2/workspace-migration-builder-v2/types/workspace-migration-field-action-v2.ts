import { FieldMetadataEntity } from 'src/engine/metadata-modules/field-metadata/field-metadata.entity';
import { FlattenFieldMetadata } from 'src/engine/workspace-manager/workspace-migration-v2/types/flat-field-metadata';
import { FlattenObjectMetadataWithoutFields } from 'src/engine/workspace-manager/workspace-migration-v2/types/flat-object-metadata';
import { FromTo } from 'src/engine/workspace-manager/workspace-migration-v2/types/from-to.type';
import { FlattenFieldMetadataPropertiesToCompare } from 'src/engine/workspace-manager/workspace-migration-v2/utils/flat-field-metadata-comparator.util';

export type FieldAndObjectMetadataWorkspaceMigrationInput = {
  flattenFieldMetadata: FlattenFieldMetadata;
  flattenObjectMetadata: FlattenObjectMetadataWithoutFields;
};
export type CreateFieldAction = {
  type: 'create_field';
} & FieldAndObjectMetadataWorkspaceMigrationInput;

export type UpdateFieldAction = {
  type: 'update_field';
  updates: Partial<
    {
      [P in FlattenFieldMetadataPropertiesToCompare]: {
        property: P;
      } & FromTo<FieldMetadataEntity[P]>;
    }[FlattenFieldMetadataPropertiesToCompare]
  >[];
} & FieldAndObjectMetadataWorkspaceMigrationInput;

export type DeleteFieldAction = {
  type: 'delete_field';
} & FieldAndObjectMetadataWorkspaceMigrationInput;

export type WorkspaceMigrationFieldActionV2 =
  | CreateFieldAction
  | UpdateFieldAction
  | DeleteFieldAction;
