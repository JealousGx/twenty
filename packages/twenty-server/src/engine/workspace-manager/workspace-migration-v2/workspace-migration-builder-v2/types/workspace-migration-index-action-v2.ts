import { type FlatIndexMetadataPropertiesToCompare } from 'src/engine/metadata-modules/flat-index-metadata/types/flat-index-metadata-properites-to-compare.type';
import { type FlatIndexMetadataPropertyUpdate } from 'src/engine/metadata-modules/flat-index-metadata/types/flat-index-metadata-property-update.type';
import { type FlatIndexMetadata } from 'src/engine/metadata-modules/flat-index-metadata/types/flat-index-metadata.type';

export type CreateIndexAction = {
  type: 'create_index';
  flatIndexMetadata: FlatIndexMetadata;
};

export type DeleteIndexAction = {
  type: 'delete_index';
  flatIndexMetadataId: string;
};

export type UpdateIndexAction = {
  type: 'update_index';
  flatIndexMetadataId: string;
  updates: Array<
    {
      [P in FlatIndexMetadataPropertiesToCompare]: FlatIndexMetadataPropertyUpdate<P>;
    }[FlatIndexMetadataPropertiesToCompare]
  >;
};

export type WorkspaceMigrationIndexActionV2 =
  | CreateIndexAction
  | DeleteIndexAction;
// | UpdateIndexAction; // TODO incoming

export type WorkspaceMigrationIndexActionTypeV2 =
  WorkspaceMigrationIndexActionV2['type'];
