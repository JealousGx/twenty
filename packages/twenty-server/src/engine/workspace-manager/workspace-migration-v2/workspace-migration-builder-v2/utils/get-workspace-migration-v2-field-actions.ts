import {
  CreateFieldAction,
  DeleteFieldAction,
  FieldAndObjectMetadataWorkspaceMigrationInput,
} from 'src/engine/workspace-manager/workspace-migration-v2/workspace-migration-builder-v2/types/workspace-migration-field-action-v2';

export const getWorkspaceMigrationV2FieldCreateAction = ({
  flattenFieldMetadata,
  flattenObjectMetadata,
}: FieldAndObjectMetadataWorkspaceMigrationInput): CreateFieldAction => ({
  type: 'create_field',
  flattenFieldMetadata,
  flattenObjectMetadata,
});

export const getWorkspaceMigrationV2FieldDeleteAction = ({
  flattenFieldMetadata,
  flattenObjectMetadata,
}: FieldAndObjectMetadataWorkspaceMigrationInput): DeleteFieldAction => ({
  type: 'delete_field',
  flattenFieldMetadata,
  flattenObjectMetadata,
});
