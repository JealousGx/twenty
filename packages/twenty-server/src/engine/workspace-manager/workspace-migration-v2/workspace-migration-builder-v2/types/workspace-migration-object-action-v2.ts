import { ObjectMetadataEntity } from 'src/engine/metadata-modules/object-metadata/object-metadata.entity';
import { FlattenObjectMetdataWithoutFields } from 'src/engine/workspace-manager/workspace-migration-v2/types/flatten-object-metadata';
import { FromTo } from 'src/engine/workspace-manager/workspace-migration-v2/types/from-to.type';
import { FlattenObjectMetadataPropertiesToCompare } from 'src/engine/workspace-manager/workspace-migration-v2/utils/flatten-object-metadata-comparator.util';

type ObjectActionCommon = {
  flattenObjectMetadata: FlattenObjectMetdataWithoutFields;
};
export type CreateObjectAction = {
  type: 'create_object';
} & ObjectActionCommon;

export type UpdateObjectAction = {
  type: 'update_object';
  updates: Partial<
    {
      [P in FlattenObjectMetadataPropertiesToCompare]: {
        property: P;
      } & FromTo<ObjectMetadataEntity[P]>;
    }[FlattenObjectMetadataPropertiesToCompare]
  >[];
} & ObjectActionCommon;

export type DeleteObjectAction = {
  type: 'delete_object';
} & ObjectActionCommon;

export type WorkspaceMigrationV2ObjectAction =
  | CreateObjectAction
  | UpdateObjectAction
  | DeleteObjectAction;
