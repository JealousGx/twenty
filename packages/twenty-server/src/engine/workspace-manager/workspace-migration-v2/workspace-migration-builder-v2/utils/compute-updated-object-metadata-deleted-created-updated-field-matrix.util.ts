import { FlattenFieldMetadata } from 'src/engine/workspace-manager/workspace-migration-v2/types/flatten-field-metadata';
import {
  FlattenObjectMetadata,
  FlattenedObjectMetdataWithoutFields,
} from 'src/engine/workspace-manager/workspace-migration-v2/types/flatten-object-metadata';
import { FromTo } from 'src/engine/workspace-manager/workspace-migration-v2/types/from-to.type';
import {
  CustomDeletedCreatedUpdatedMatrix,
  deletedCreatedUpdatedMatrixDispatcher,
} from 'src/engine/workspace-manager/workspace-migration-v2/workspace-migration-builder-v2/utils/deleted-created-updated-matrix-dispatcher.util';

export type UpdatedObjectMetadataDeletedCreatedUpdatedFieldMatrix = {
  objectMetadataInput: FlattenedObjectMetdataWithoutFields;
} & CustomDeletedCreatedUpdatedMatrix<'fieldMetadata', FlattenFieldMetadata>;

export const computeUpdatedObjectMetadataDeletedCreatedUpdatedFieldMatrix = (
  updatedObjectMetadata: FromTo<FlattenObjectMetadata>[],
): UpdatedObjectMetadataDeletedCreatedUpdatedFieldMatrix[] => {
  const matrixAccumulator: UpdatedObjectMetadataDeletedCreatedUpdatedFieldMatrix[] =
    [];

  for (const { from, to } of updatedObjectMetadata) {
    const fieldMetadataMatrix = deletedCreatedUpdatedMatrixDispatcher({
      from: from.fieldInputs,
      to: to.fieldInputs,
    });

    matrixAccumulator.push({
      objectMetadataInput: to,
      createdFieldMetadata: fieldMetadataMatrix.created,
      deletedFieldMetadata: fieldMetadataMatrix.deleted,
      updatedFieldMetadata: fieldMetadataMatrix.updated,
    });
  }

  return matrixAccumulator;
};
