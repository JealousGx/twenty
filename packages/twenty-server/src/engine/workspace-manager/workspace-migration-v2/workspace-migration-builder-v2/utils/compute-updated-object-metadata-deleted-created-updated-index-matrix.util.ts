import { FromTo } from 'twenty-shared/types';

import {
  FlatObjectMetadata,
  FlatObjectMetadataWithoutFields,
} from 'src/engine/metadata-modules/flat-object-metadata/types/flat-object-metadata';
import { FlatIndexMetadata } from 'src/engine/workspace-manager/workspace-migration-v2/types/flat-index-metadata';
import {
  CustomDeletedCreatedUpdatedMatrix,
  deletedCreatedUpdatedMatrixDispatcher,
} from 'src/engine/workspace-manager/workspace-migration-v2/utils/deleted-created-updated-matrix-dispatcher.util';

export type UpdatedObjectMetadataDeletedCreatedUpdatedIndexMatrix = {
  flatObjectMetadata: FlatObjectMetadataWithoutFields;
} & CustomDeletedCreatedUpdatedMatrix<'indexMetadata', FlatIndexMetadata>;

export const computeUpdatedObjectMetadataDeletedCreatedUpdatedIndexMatrix = (
  updatedObjectMetadata: FromTo<FlatObjectMetadata>[],
): UpdatedObjectMetadataDeletedCreatedUpdatedIndexMatrix[] => {
  const matrixAccumulator: UpdatedObjectMetadataDeletedCreatedUpdatedIndexMatrix[] =
    [];

  for (const { from, to } of updatedObjectMetadata) {
    const indexMetadataMatrix = deletedCreatedUpdatedMatrixDispatcher({
      from: from.flatIndexMetadatas,
      to: to.flatIndexMetadatas,
    });

    matrixAccumulator.push({
      flatObjectMetadata: to,
      createdIndexMetadata: indexMetadataMatrix.created,
      deletedIndexMetadata: indexMetadataMatrix.deleted,
      updatedIndexMetadata: indexMetadataMatrix.updated,
    });
  }

  return matrixAccumulator;
};
