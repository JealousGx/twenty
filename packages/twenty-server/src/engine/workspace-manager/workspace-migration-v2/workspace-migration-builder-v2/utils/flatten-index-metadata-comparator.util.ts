import diff from 'microdiff';

import { FlattenIndexMetadata } from 'src/engine/workspace-manager/workspace-migration-v2/types/flatten-index-metadata';
import { FromTo } from 'src/engine/workspace-manager/workspace-migration-v2/types/from-to.type';
import { transformMetadataForComparison } from 'src/engine/workspace-manager/workspace-sync-metadata/comparators/utils/transform-metadata-for-comparison.util';

const flattenedIndexMetadataPropertiesToCompare = [
  'flattenIndexFieldMetadatas', // Comparing this as whole ? should iterate on each keys ? => TBD should only map over cols as before ?
  'indexType',
  'indexWhereClause',
  'isUnique',
  'name',
] as const satisfies (keyof FlattenIndexMetadata)[];
type FlattenedIndexMetadataPropertiesToCompare =
  (typeof flattenedIndexMetadataPropertiesToCompare)[number];

// Should also handle indexFieldMetadata comparison ?
export const compareTwoFlattenedIndexMetadata = ({
  from,
  to,
}: FromTo<FlattenIndexMetadata>) => {
  const transformOptions = {
    shouldIgnoreProperty: (property: string) =>
      !flattenedIndexMetadataPropertiesToCompare.includes(
        property as FlattenedIndexMetadataPropertiesToCompare,
      ),
  };

  const fromCompare = transformMetadataForComparison(from, transformOptions);
  const toCompare = transformMetadataForComparison(to, transformOptions);

  const indexesDifferences = diff(fromCompare, toCompare);

  return indexesDifferences.flatMap<{ property: string } & FromTo<unknown>>(
    (difference) => {
      switch (difference.type) {
        case 'CHANGE': {
          const { oldValue, path, value } = difference;

          const property = path[0];

          if (typeof property === 'number') {
            return [];
          }

          return {
            from: oldValue,
            property,
            to: value,
          };
        }
        case 'CREATE':
        case 'REMOVE':
        default:
          return [];
      }
    },
  );
};
