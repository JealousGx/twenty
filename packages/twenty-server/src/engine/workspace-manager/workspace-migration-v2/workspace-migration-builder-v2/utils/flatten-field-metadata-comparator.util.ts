import diff from 'microdiff';
import { FieldMetadataType } from 'twenty-shared/types';
import { isDefined } from 'twenty-shared/utils';

import { FlattenFieldMetadata } from 'src/engine/workspace-manager/workspace-migration-v2/types/flatten-field-metadata';
import { FromTo } from 'src/engine/workspace-manager/workspace-migration-v2/types/from-to.type';
import { UpdateFieldAction } from 'src/engine/workspace-manager/workspace-migration-v2/types/workspace-migration-field-action-v2';
import { transformMetadataForComparison } from 'src/engine/workspace-manager/workspace-sync-metadata/comparators/utils/transform-metadata-for-comparison.util';

const workspaceMigrationFieldInputPropertiesToCompare = [
  'defaultValue',
  'description',
  'icon',
  'isActive',
  'isLabelSyncedWithName',
  'isUnique',
  'label',
  'name',
  'options',
  'standardOverrides',
] as const satisfies (keyof FlattenFieldMetadata)[];
export type WorkspaceMigrationFieldInputPropertiesToCompare =
  (typeof workspaceMigrationFieldInputPropertiesToCompare)[number];

const fieldMetadataPropertiesToStringify = [
  'defaultValue',
  'standardOverrides',
] as const satisfies WorkspaceMigrationFieldInputPropertiesToCompare[];

const shouldNotOverrideDefaultValue = (type: FieldMetadataType) => {
  return [
    FieldMetadataType.BOOLEAN,
    FieldMetadataType.SELECT,
    FieldMetadataType.MULTI_SELECT,
    FieldMetadataType.CURRENCY,
    FieldMetadataType.PHONES,
    FieldMetadataType.ADDRESS,
  ].includes(type);
};

type GetWorkspaceMigrationUpdateFieldActionArgs = FromTo<FlattenFieldMetadata>;
export const compareTwoFlattenFieldMetadata = ({
  from,
  to,
}: GetWorkspaceMigrationUpdateFieldActionArgs) => {
  const compareFieldMetadataOptions = {
    shouldIgnoreProperty: (
      property: string,
      fieldMetadata: FlattenFieldMetadata,
    ) => {
      if (
        !workspaceMigrationFieldInputPropertiesToCompare.includes(
          property as WorkspaceMigrationFieldInputPropertiesToCompare,
        )
      ) {
        return true;
      }

      if (
        property === 'defaultValue' &&
        isDefined(fieldMetadata.type) &&
        shouldNotOverrideDefaultValue(fieldMetadata.type)
      ) {
        return true;
      }

      return false;
    },
    propertiesToStringify: fieldMetadataPropertiesToStringify,
  };
  const fromCompare = transformMetadataForComparison(
    from,
    compareFieldMetadataOptions,
  );
  const toCompare = transformMetadataForComparison(
    to,
    compareFieldMetadataOptions,
  );

  const flattenFieldMetadataDifferences = diff(fromCompare, toCompare);

  return flattenFieldMetadataDifferences.flatMap<
    UpdateFieldAction['updates'][number]
  >((difference) => {
    switch (difference.type) {
      case 'CHANGE': {
        const { oldValue, path, value } = difference;

        return {
          from: oldValue,
          to: value,
          property: path[0] as WorkspaceMigrationFieldInputPropertiesToCompare,
        };
      }
      case 'CREATE':
      case 'REMOVE':
      default: {
        // Should never occurs, we should only provide null never undefined and so on
        return [];
      }
    }
  });
};
