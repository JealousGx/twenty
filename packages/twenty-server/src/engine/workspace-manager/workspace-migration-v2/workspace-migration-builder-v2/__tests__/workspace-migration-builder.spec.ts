import { WORKSPACE_MIGRATION_OBJECT_BUILDER_TEST_CASES } from 'src/engine/workspace-manager/workspace-migration-v2/workspace-migration-builder-v2/__tests__/common/workspace-migration-builder-object-rest-case';
import {
  CamelCasedWorkspaceMigrationActionsType,
  ExpectedActionCounters,
} from 'src/engine/workspace-manager/workspace-migration-v2/workspace-migration-builder-v2/__tests__/common/workspace-migration-builder-test-case.type';
import { WorkspaceMigrationV2 } from 'src/engine/workspace-manager/workspace-migration-v2/workspace-migration-builder-v2/types/workspace-migration-v2';
import { WorkspaceMigrationBuilderV2Service } from 'src/engine/workspace-manager/workspace-migration-v2/workspace-migration-builder-v2/workspace-migration-builder-v2.service';
import { extractRecordIdsAndDatesAsExpectAny } from 'test/utils/extract-record-ids-and-dates-as-expect-any';
import { capitalize } from 'twenty-shared/utils';

describe('Workspace migration builder object actions test suite', () => {
  let service: WorkspaceMigrationBuilderV2Service;

  beforeEach(() => {
    service = new WorkspaceMigrationBuilderV2Service();
  });

  const expectedActionsTypeCounterChecker = ({
    expectedActionsTypeCounter,
    workspaceMigration,
  }: {
    workspaceMigration: WorkspaceMigrationV2;
    expectedActionsTypeCounter: ExpectedActionCounters;
  }) => {
    const initialAcc: ExpectedActionCounters = { total: 0 };
    const actualActionsTypeCounter =
      workspaceMigration.actions.reduce<ExpectedActionCounters>(
        (acc, action) => {
          const { type } = action;
          const [operation, target] = type.split('_');
          const formattedActionKey =
            `${operation}${capitalize(target)}` as CamelCasedWorkspaceMigrationActionsType;

          return {
            ...acc,
            total: acc.total + 1,
            [formattedActionKey]: acc[formattedActionKey] ?? 0 + 1,
          };
        },
        initialAcc,
      );

    expect(actualActionsTypeCounter).toEqual(expectedActionsTypeCounter);
  };

  it.each(WORKSPACE_MIGRATION_OBJECT_BUILDER_TEST_CASES)(
    '$title',
    ({ context: { input, expectedActionsTypeCounter } }) => {
      const { from, to } = typeof input === 'function' ? input() : input;
      const workspaceMigration = service.build({
        from,
        to,
      });

      expectedActionsTypeCounterChecker({
        expectedActionsTypeCounter,
        workspaceMigration,
      });
      const { actions, ...rest } = workspaceMigration;
      expect(actions).toMatchSnapshot(
        actions.map(extractRecordIdsAndDatesAsExpectAny),
      );
      expect(rest).toMatchSnapshot();
    },
  );
});
