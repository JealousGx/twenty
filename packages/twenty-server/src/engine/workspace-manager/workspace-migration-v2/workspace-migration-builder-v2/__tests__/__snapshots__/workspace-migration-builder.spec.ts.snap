// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workspace migration builder field actions test suite It should build an create_field action 1`] = `
[
  {
    "flattenFieldMetadata": {
      "createdAt": Any<ClockDate>,
      "defaultValue": undefined,
      "description": "default flatten field metadata description",
      "icon": "icon",
      "id": Any<String>,
      "isActive": true,
      "isCustom": true,
      "isLabelSyncedWithName": false,
      "isNullable": true,
      "isSystem": false,
      "isUnique": false,
      "label": "flatten field metadata label",
      "name": "flattenFieldMetadataName",
      "objectMetadataId": Any<String>,
      "options": undefined,
      "relationTargetFieldMetadata": undefined,
      "relationTargetFieldMetadataId": undefined,
      "relationTargetObjectMetadata": undefined,
      "relationTargetObjectMetadataId": undefined,
      "settings": undefined,
      "standardId": undefined,
      "standardOverrides": undefined,
      "type": "TEXT",
      "uniqueIdentifier": "poire",
      "updatedAt": Any<ClockDate>,
      "workspaceId": Any<String>,
    },
    "flattenObjectMetadata": {
      "createdAt": Any<ClockDate>,
      "dataSourceId": Any<String>,
      "description": "default flatten object metadata description",
      "duplicateCriteria": [],
      "flattenFieldMetadatas": [
        {
          "createdAt": Any<ClockDate>,
          "defaultValue": undefined,
          "description": "default flatten field metadata description",
          "icon": "icon",
          "id": Any<String>,
          "isActive": true,
          "isCustom": true,
          "isLabelSyncedWithName": false,
          "isNullable": true,
          "isSystem": false,
          "isUnique": false,
          "label": "flatten field metadata label",
          "name": "flattenFieldMetadataName",
          "objectMetadataId": Any<String>,
          "options": undefined,
          "relationTargetFieldMetadata": undefined,
          "relationTargetFieldMetadataId": undefined,
          "relationTargetObjectMetadata": undefined,
          "relationTargetObjectMetadataId": undefined,
          "settings": undefined,
          "standardId": undefined,
          "standardOverrides": undefined,
          "type": "TEXT",
          "uniqueIdentifier": "poire",
          "updatedAt": Any<ClockDate>,
          "workspaceId": Any<String>,
        },
      ],
      "flattenIndexMetadatas": [],
      "icon": "icon",
      "id": Any<String>,
      "imageIdentifierFieldMetadataId": Any<String>,
      "isActive": true,
      "isAuditLogged": true,
      "isCustom": true,
      "isLabelSyncedWithName": true,
      "isRemote": false,
      "isSearchable": true,
      "isSystem": false,
      "labelIdentifierFieldMetadataId": Any<String>,
      "labelPlural": "default flatten object metadata label plural",
      "labelSingular": "default flatten object metadata label singular",
      "namePlural": "defaultFlattenObjectMetdataNamePlural",
      "nameSingular": "defaultFlattenObjectMetdataNameSingular",
      "shortcut": "shortcut",
      "standardId": undefined,
      "standardOverrides": undefined,
      "targetRelationFields": [],
      "targetTableName": "",
      "uniqueIdentifier": "pomme",
      "updatedAt": Any<ClockDate>,
      "workspaceId": Any<String>,
    },
    "type": "create_field",
  },
]
`;

exports[`Workspace migration builder field actions test suite It should build an create_field action 2`] = `{}`;
